source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -158,55 +170,159 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
       // If the request is forbidden by the default token bucket, return the result and the
       // default token bucket.
-      return {false, makeOptRefFromPtr<TokenBucketContext>(default_token_bucket_.get())};
+      return {false, std::shared_ptr<TokenBucketContext>(default_token_bucket_)};
     }
 
     // If the request is allowed then return the result the token bucket. The descriptor
     // token bucket will be selected as priority if it exists.
-    return {true, makeOptRef<TokenBucketContext>(matched_results.empty()
-                                                     ? *default_token_bucket_
-                                                     : matched_results[0].token_bucket.get())};
+    return {true,
+            matched_results.empty() ? default_token_bucket_ : matched_results[0].token_bucket};
   };
 
   ASSERT(!matched_results.empty());
-  return {true, makeOptRef<TokenBucketContext>(matched_results[0].token_bucket.get())};
+  std::shared_ptr<TokenBucketContext> bucket_context =
+      std::shared_ptr<TokenBucketContext>(matched_results[0].token_bucket);
+  return {true, bucket_context};
+}
+
+// Compare the request descriptor entries with the user descriptor entries. If all non-empty user
+// descriptor values match the request descriptor values, return true
+bool DynamicDescriptorMap::matchDescriptorEntries(
+    const std::vector<RateLimit::DescriptorEntry>& request_entries,
+    const std::vector<RateLimit::DescriptorEntry>& config_entries) {
+  // Check for equality of sizes
+  if (request_entries.size() != config_entries.size()) {
+    return false;
+  }
+
+  for (size_t i = 0; i < request_entries.size(); ++i) {
+    // Check if the keys are equal.
+    if (request_entries[i].key_ != config_entries[i].key_) {
+      return false;
+    }
+
+    // Check values are equal or wildcard value is used.
+    if (config_entries[i].value_.empty()) {
+      continue;
+    }
+    if (request_entries[i].value_ != config_entries[i].value_) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void DynamicDescriptorMap::addDescriptor(const RateLimit::LocalDescriptor& user_descriptor,
+                                         DynamicDescriptorSharedPtr dynamic_descriptor) {
+  auto result = user_descriptors_.emplace(user_descriptor, std::move(dynamic_descriptor));
+  if (!result.second) {
+    throw EnvoyException(absl::StrCat("duplicate descriptor in the local rate descriptor: ",
+                                      result.first->first.toString()));
+  }
+}
+
+RateLimitTokenBucketSharedPtr
+DynamicDescriptorMap::getBucket(const RateLimit::Descriptor request_descriptor) {
+  for (const auto& pair : user_descriptors_) {
+    auto user_descriptor = pair.first;
+    if (!matchDescriptorEntries(request_descriptor.entries_, user_descriptor.entries_)) {
+      continue;
+    }
+
+    // we found a user configured wildcard descriptor that matches the request descriptor.
```

Comment:
Please all check all these comments match our style.

Reply:
I have updated this specific comment. Is there a doc/readme that have guidelines about the comment styles for my reference?